File: src/main/java/com/back/together02be/trade/entity/Trade.java
package com.back.together02be.trade.entity;

import java.time.LocalDateTime;

import com.back.together02be.global.entity.BaseEntity;
import com.back.together02be.stock.entity.Stock;
import com.back.together02be.users.entity.Users;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor
public class Trade extends BaseEntity {

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "users_id", nullable = false)
	private Users users;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "stock_id", nullable = false)
	private Stock stock;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private TradeType type;

	@Column(nullable = false)
	private Long quantity;

	@Column(nullable = false)
	private Long price;

	@Column(nullable = false)
	private Long amount;

	private Long profit;

	@Column(name = "traded_at", nullable = false)
	private LocalDateTime tradedAt;

	public Trade(Users users, Stock stock, TradeType type, Long quantity, Long price, Long profit) {
		this.users = users;
		this.stock = stock;
		this.type = type;
		this.quantity = quantity;
		this.price = price;
		this.amount = (long) quantity * price;
		this.profit = profit;
		this.tradedAt = LocalDateTime.now();
	}

	// 매수: profit은 무조건 null
	public static Trade buy(Users users, Stock stock,
		Long quantity, Long price) {
		return new Trade(users, stock, TradeType.BUY, quantity, price, null);
	}

	// 매도: profit 필수
	public static Trade sell(Users users, Stock stock,
		Long quantity, Long price, Long profit) {
		return new Trade(users, stock, TradeType.SELL, quantity, price, profit);
	}
}

File: src/main/java/com/back/together02be/trade/entity/Trade.kt
package com.back.together02be.trade.entity

import com.back.together02be.global.entity.BaseEntity
import com.back.together02be.stock.entity.Stock
import com.back.together02be.users.entity.Users
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import java.time.LocalDateTime

@Entity
class Trade private constructor(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "users_id", nullable = false)
    val users: Users,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "stock_id", nullable = false)
    val stock: Stock,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val type: TradeType,

    @Column(nullable = false)
    val quantity: Long,

    @Column(nullable = false)
    val price: Long,

    @Column(nullable = false)
    val amount: Long = quantity * price,

    val profit: Long? = null,

    @Column(name = "traded_at", nullable = false)
    val tradedAt: LocalDateTime = LocalDateTime.now()
) : BaseEntity() {

    // JPA가 프록시 객체를 생성할 때 필요한 기본 생성자 (Protected 수준)
    protected constructor() : this(
        users = null!!,
        stock = null!!,
        type = TradeType.BUY,
        quantity = 0L,
        price = 0L,
        amount = 0L,
        profit = null,
        tradedAt = LocalDateTime.now()
    )

    companion object {
        // 매수: profit은 무조건 null
        fun buy(users: Users, stock: Stock, quantity: Long, price: Long): Trade {
            return Trade(
                users = users,
                stock = stock,
                type = TradeType.BUY,
                quantity = quantity,
                price = price,
                profit = null
            )
        }

        // 매도: profit 필수
        fun sell(users: Users, stock: Stock, quantity: Long, price: Long, profit: Long): Trade {
            return Trade(
                users = users,
                stock = stock,
                type = TradeType.SELL,
                quantity = quantity,
                price = price,
                profit = profit
            )
        }
    }
}